### PR TITLE
1261153 - Added disk/ram/cpu validations to fusor-installer.

### PR DESCRIPTION
**Pull request by djwhatle**
Added system hardware validations to fusor-installer. disk/ram/cpu will all be checked against required and recommended quantities. Required and recommended quantities taken from QCI docs, with exception of disk space. Disk space was found to exceed requirements given in docs (already submitted a BZ)

/ and /var/lib (eventually containing /var/lib/pulp) are checked for capacity.

User has --skip_validations option if system requirement checking if not needed.

### CODE REVIEW

File: bin/fusor-installer [hunk 1]
```diff
@@ -1,90 +1,226 @@
...
-  while [[ $# > 0 ]]; do
-    i="$1"
-    case $i in
-      --deployment_dir=*)
-      deployment_dir="${i#*=}"
-      shift
-      ;;
-      --deployment_dir)
-      deployment_dir="$2"
+fusor_install() {
```
Comment (jmrodri) [inline bin/fusor-installer:3]: Putting it in a function is a GREAT choice :) I thought this was already done but apparently that was the undercloud-installer.

**Comment by jmrodri** [comment]
Man the diff is horrible because we put it in a function it can't tell me what was there vs what was really added. I was hoping for a quick review. I'll take a closer look a little later.

File: bin/fusor-installer [hunk 2]
```diff
@@ -1,90 +1,226 @@
...
+  if [ "$(id -u)" != "0" ]; then
+     echo "This script must be run as root"
+     exit 1
+  fi
+
+  interactive=true
```
Comment (jmrodri) [inline bin/fusor-installer:9]: Where do we unset `interactive`? We set it to true here and check it 3 times below but never do we set it to false that I can see.

File: bin/fusor-installer [hunk 3]
```diff
@@ -1,90 +1,226 @@
...
+  # Check if --skip_validations option was supplied
+  if [[ $@ != *"--skip_validations"* ]]; then
+    echo
+    # 2 cores required, 4 recommended
+    check_cpu_cores 2 4
+    # 16 gb ram required, 16 gb recommended
+    check_ram 16 16
+    # 200 gb disk space at '/' required, 250 gb recommended
+    check_disk_space 200 250 "/"
+    # 200 gb disk space at '/' required, 250 gb recommended
```
Comment (jmrodri) [inline bin/fusor-installer]: This comment doesn't match the directory you are passing below.

**Comment by jmrodri** [comment]
I noticed that `check_cpu_cores`, `check_ram` and `check_disk_space` are identical save for 2 items: how you get the found value i.e. `dmidecode, nprocs` and the prompt for interactive mode. I probably would've gone one step further in creating another function that did all the comparison that took all the missing params

``` bash
comparison(intro_msg, found_val, int_prompt, required, recommended)
  # do all the -ge checks here 
```

Then your check functions become:

``` bash
check_cpu_cores() {
  # Check if machine has enough CPU cores
  cores_required=$1
  cores_recommended=$2

  intro_message="Checking CPU cores..."
  cores_found=$(nproc)
  prompt = "Proceed without meeting CPU core requirement? (y/n) "

  comparison(intro_message, cores_found, prompt, cores_required, cores_recommended)
}
```

Something like that. Not 100% sure if it will make it easier or just confuses the situation.

File: bin/fusor-installer (hunk 2 above)
Comment (djwhatle) [inline bin/fusor-installer:9]: I wrote the code in fusor-undercloud-installer first, then transplanted it over to fusor-installer.

In an effort to have functions check_cpu_cores etc remain exactly the same, I included an interactive variable in this script as well.

Although interactive is presently the only mode available for fusor-installer,  I reasoned that we may implement a non-interactive mode down the road.

**Comment by jmrodri** [comment]
Hrm I saw more differences. Damn they're just ever so slightly to mess up reusing the code. The `print_with_outcome` is passed in MB, cores, or GB depending on which check it is :(

File: bin/fusor-installer (hunk 3 above)
Comment (djwhatle) [inline bin/fusor-installer]: Thanks. Fixed.

**Comment by djwhatle** [comment]
I'll take a stab at generalizing the code some more. Thanks for the suggestions.

**Comment by jmrodri** [comment]
Don't bother, it's probably more work than needed.

**Comment by jmrodri** [comment]
ACK

**Pull request merged**